Preserve node's stack trace during retrace

SherlockNoMad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83050

AOTAutograd retraces graph module produced by torch dynamo, this PR preserves the stack trace in the original fx.Node.

Differential Revision: [D38595638](https://our.internmc.facebook.com/intern/diff/D38595638)

SherlockNoMad committed:
Preserve node's stack trace during retrace

[ghstack-poisoned]

SherlockNoMad commented on torch/fx/proxy.py:
```diff
@@ -42,7 +42,13 @@ def create_node(self, kind : str, target : Target,
             check_for_mutable_operation(target, args, kwargs)
 
-        return self.graph.create_node(kind, target, args, kwargs, name, type_expr)
+        node = self.graph.create_node(kind, target, args, kwargs, name, type_expr)
```
Or shall I only apply this behavior to PythonKeyTracer? 
Do we want PythonKeyTracer to be the parents class for all retracing retracer ?

Chillee replied:
I think we can probably put this logic inside of ProxyTensor itself, right?

SherlockNoMad replied:
I am not sure how to do it... or if it's possible? 

To propagate stack trace from one FX node to another FX node in the new graph, I am reading stack trace from an Interpreter.current_node, and writing stack trace to a new node being created by a Tracer. 

IIUC, ProxyTensor and ProxyTorchDispatchMode doesn't carry such stack_trace info...

SherlockNoMad commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -382,6 +383,24 @@ def wrapped(*args):
 
     return wrapped
 
+class RetracingMode:
```
I know this is not ideal... and I hate to save the interpreter in a global variable. 

but AOTAutograd wraps gm.forward with multiple layers of wrapper function, and by the time it reaches the make_fx function, there is no good way to access the binded interpreter anymore. 

Is there a better way to do this?

Chillee replied:
hrmmm.... I think it's acceptable. Perhaps we should put this info on the ProxyTensorMode instead?

SherlockNoMad replied:
I suppose you mean ProxyTorchDispatchMode... and I am not sure how to do that
We can chat offline if you think it's still doable that way.

ezyang reviewed (approved):
This seems fine. You should write a test. Note that I'd characterize this as "overriding" the stack trace that would normally be collected; make sure they're not colliding if you also have the "save stack traces" bool set!

SherlockNoMad committed:
Update on "Preserve node's stack trace during retrace"


AOTAutograd retraces graph module produced by torch dynamo, this PR preserves the stack trace in the original fx.Node. 

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Preserve node's stack trace during retrace"


AOTAutograd retraces graph module produced by torch dynamo, this PR preserves the stack trace in the original fx.Node. 

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Preserve node's stack trace during retrace"


AOTAutograd retraces graph module produced by torch dynamo, this PR preserves the stack trace in the original fx.Node. 

[ghstack-poisoned]

SherlockNoMad committed:
Update on "Preserve node's stack trace during retrace"


AOTAutograd retraces graph module produced by torch dynamo, this PR preserves the stack trace in the original fx.Node. 

[ghstack-poisoned]

Chillee commented on torch/fx/traceback.py:
```diff
@@ -0,0 +1,35 @@
+    global active_interpreter
+
+    if active_interpreter:
+        node = getattr(active_interpreter, "current_node", None)
```
If `active_interpreter` is None, don't we error on the following line anyways?

SherlockNoMad committed:
Update on "Preserve node's stack trace during retrace"


AOTAutograd retraces graph module produced by torch dynamo, this PR preserves the stack trace in the original fx.Node. 

[ghstack-poisoned]

ezyang commented on functorch/functorch/_src/aot_autograd.py:
```diff
@@ -657,7 +658,13 @@ def functional_call(*args, **kwargs):
             mod, pytree.tree_unflatten(args[:params_len], params_spec)
         ):
-            out = mod(*args[params_len:], **kwargs)
+            if isinstance(mod, torch.fx.GraphModule):
```
I won't block this, but it will be better if the call sites take care of interpreter-ing the GraphModule themselves. By doing so, they will ensure that this works even if you have a GraphModule and you wrap it in some dummy adapter function to massage the input/outputs. We do this in several cases. This logic won't work for that case.

ezyang replied:
Even with this API, it is still relatively easy to accidentally do the wrong thing and lose stack traces. If we can't design the API so that this is impossible to happen (and given that I nuked your "change the compiled code" change, I don't think such an API design is possible), it is better to catch it through testing/linting and force people to think about it explicitly.

ezyang commented on torch/fx/traceback.py:
```diff
@@ -0,0 +1,40 @@
+__all__ = ['override_stack_trace', 'current_stack_trace', 'is_stack_trace_overridden']
+
+active_interpreter = None
+is_overriden = False
```
overridden

ezyang replied:
It seems like is_overridden is always True when active interpreter is not None, so you don't need both variables?

ezyang commented on torch/fx/traceback.py:
```diff
@@ -0,0 +1,40 @@
+    global active_interpreter
+    global is_overriden
+
+    saved_intepreter = active_interpreter
```
interpreter

ezyang commented:
You're currently storing the dynamic scope on Interpreter. I think it may make more sense to store it on the Tracer. The reasoning here is that the Tracer is what actually is going to store the traceback on the object, so logically if you want to modify what traceback the Tracer is going to store, you ought to be monkeying around with the Tracer's state. This is feasible for the Interpreter to do if you pass the Tracer to the Interpreter, and although in the current version of this PR you don't have a Tracer handy when you set the scope, you DO have it handy at the test site, and I suggested you have the test site be responsible for arming the Interpreter with the Tracer object to set stack traces on. Then you just need a context manager on the tracer for managing the dynamic scoping. WDYT

albanD commented on torch/fx/traceback.py:
```diff
@@ -0,0 +1,40 @@
+        is_overriden = saved_is_overriden
+
+@compatibility(is_backward_compatible=False)
+def current_stack_trace() -> Optional[str]:
```
Could we make this the exact same API as `traceback.format_stack()` ? That way, we can simply replace `traceback.format_stack()` calls with `torch.fx.format_stack()` in our code and it will just work (I'm thinking about anomaly mode in particular).

SherlockNoMad replied:
Sure, done.

SherlockNoMad committed:
Update on "Preserve node's stack trace during retrace"


AOTAutograd retraces graph module produced by torch dynamo, this PR preserves the stack trace in the original fx.Node. 

[ghstack-poisoned]

ezyang commented on torch/fx/traceback.py:
```diff
@@ -0,0 +1,54 @@
+
+__all__ = ['override_stack_trace', 'append_stack_trace', 'format_stack', 'is_stack_trace_overridden']
+
+current_stack: List[str] = []
```
there seems to be some stack confusion here: is this a list of lines of a stacktrace, or a list of entire stacktraces. Looks like the former, from reading the code

SherlockNoMad replied:
this is stack of stacks, the latter one.

ezyang replied:
This doesn't seem consistent with the `append_stack_trace` then, which seems to push an entire stack trace (all lines)

SherlockNoMad replied:
OK, I was confused by "list of lines of a stacktrace" and "list of entire stacktraces"... 

the content of stack variable here is something like this 

```
  File "/fsx/users/bahuang/repos/pytorch_fsx/test_traceback.py", line 27, in <module>
    f(1)
  File "/fsx/users/bahuang/repos/pytorch_fsx/test_traceback.py", line 6, in f
    return ff(a)
  File "/fsx/users/bahuang/repos/pytorch_fsx/test_traceback.py", line 9, in ff
    return fff(a)
  File "/fsx/users/bahuang/repos/pytorch_fsx/test_traceback.py", line 12, in fff
    stack = traceback.print_stack()
```

I suppose this the "entire stack trace (all lines)"

albanD commented on torch/fx/traceback.py:
```diff
@@ -0,0 +1,54 @@
+
+@compatibility(is_backward_compatible=False)
+def is_stack_trace_overridden() -> bool:
+    global is_overridden
```
no need for global if you just read it.
Same for other functions above

albanD commented on torch/fx/traceback.py:
```diff
@@ -0,0 +1,54 @@
+    global current_stack
+
+    if is_overridden:
+        return '\n'.join(reversed(current_stack))
```
Note that `traceback.format_stack` returns the iterable, not a full string.

SherlockNoMad replied:
good catch.

albanD commented on torch/fx/traceback.py:
```diff
@@ -0,0 +1,54 @@
+    if is_overridden:
+        return '\n'.join(reversed(current_stack))
+    else:
+        return ''
```
Why isn't this `return traceback.format_stack()` ?

SherlockNoMad replied:
Done.

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38595638).

SherlockNoMad committed:
Update on "Preserve node's stack trace during retrace"


AOTAutograd retraces graph module produced by torch dynamo, this PR preserves the stack trace in the original fx.Node.

Differential Revision: [D38595638](https://our.internmc.facebook.com/intern/diff/D38595638)

[ghstack-poisoned]

ezyang commented on functorch/test/test_pythonkey.py:
```diff
@@ -547,6 +547,41 @@ def forward(self, x, y):
+        tracer = torch.fx.Tracer()
+        tracer.record_stack_traces = True
+        graph = tracer.trace(MockModule())
+        mod = torch.fx.GraphModule(tracer.root, graph)
```
This call pattern doesn't look very realistic; aren't most people going to be calling `make_fx` in PT2?

SherlockNoMad replied:
Yes, I know... this is the easiest way to populate the stack trace for a graph module ... 

make_fx is another entry point where graph module can get created... Do we want to introduce a flag to enable record_stacktrace for pythonkey_tracer within make_fx?

SherlockNoMad commented:
@pytorchbot merge

SherlockNoMad commented:
@pytorchbot merge -g

SherlockNoMad commented:
@pytorchbot merge -f "internal diff has been unlinked"

SherlockNoMad commented:
@SherlockNoMad has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38595638).

SherlockNoMad commented:
@pytorchbot merge
